evm/src/generation/mod.rs

```diff
@@ -191,6 +207,21 @@ fn apply_metadata_and_tries_memops<F: RichField + Extendable<D>, const D: usize>
     state.traces.memory_ops.extend(ops);
 }
 
+fn preinitialize_mpts<F: RichField + Extendable<D>, const D: usize>(
+    state: &mut GenerationState<F>,
+    trie_inputs: &TrieInputs,
+) -> (TrieRootPtrs, usize) {
+    let (trie_roots_ptrs, trie_data) =
+        load_all_mpts(trie_inputs).expect("Invalid MPT data for preinitialization");
+
+    for (i, data) in trie_data.iter().enumerate() {
+        let trie_addr = MemoryAddress::new(0, Segment::TrieData, i);
+        state.memory.set(trie_addr, data.into());
+    }
```

Comment:
nit: similarly to how the Kernel code is initialized, it may be a tad cleaner to do

```suggestion
        let trie_data_u256s = trie_data.iter().map(|&x| x.into()).collect();
        state.memory.contexts[0].segments[Segment::TrieData as usize].content = trie_data_u256s;
```